File: daily/2025/2381_Shifting_Letters_II.cpp
#include <iostream>
#include <vector>
#include <string>

using namespace std;

class Solution {
public:
    string shiftingLetters(string s, vector<vector<int>>& shifts) {
        int n = s.length();
        vector<int> d(n+1, 0);
        for (auto v : shifts){
            int start = v[0];
            int end = v[1];
            int dir = (v[2] == 0)? -1 : 1;
            d[start] += dir;
            d[end+1] -= dir;
        }
        int acc = 0;
        for (int i = 0; i < n; i++){
            acc += d[i];
            int curr = s[i] - 'a';
            curr = ((curr + acc) % 26 + 26) % 26;
            s[i] = 'a' + curr;
        }
        return s;
    }
};

int main(){
    string s = "abc";
    vector<vector<int>> shifts = {{0,1,0},{1,2,1},{0,2,1}};
    // string s = "dztz";
    // vector<vector<int>> shifts = {{0,0,0},{1,1,1}};
    Solution S;
    cout << S.shiftingLetters(s, shifts) << endl;
    return 0;
}

File: daily_python/2381_Shifting_Letters_II.py
import os
import sys
from typing import List

class Solution:
    def shiftingLetters(self, s: str, shifts: List[List[int]]) -> str:
        n = len(s)
        d = [0] * (n + 1)
        for start, end, dir in shifts:
            dir = 1 if dir == 1 else -1
            d[start] += dir
            d[end+1] -= dir
        acc = 0
        ans = []
        for i in range(n):
            acc += d[i]
            curr = (ord(s[i]) - ord('a') + acc) % 26
            ans.append(chr(curr + ord('a')))
        return "".join(ans)
    
def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read().strip()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            s = lines[0].split('=')[1].strip()[1:-2]
            shifts = [list(map(int, i.split(','))) for i in lines[1].split('=')[1].strip()[2:-3].split("},{")]
            dataset.append((s, shifts))
    return dataset

if __name__ == '__main__':
    if len(sys.argv) == 1:
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for s, shifts in dataset:
        results.append(solution.shiftingLetters(s, shifts))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')